Scratch PR for comments

Ericson2314 opened the pull request.

Ericson2314 committed:
Copy Template

Ericson2314 committed:
First version for opening PR

Co-authored-by: Ryan Mulligan <ryan@ryantm.com>
Co-authored-by: Connor Brewster <cbrewster@hey.com>
Co-authored-by: Ben Radford <benradf@users.noreply.github.com>
Co-authored-by: Divam <dfordivam@protonmail.com>

ryantm commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,197 @@
+# Motivation
+[motivation]: #motivation
+
+TODO: link Replit blog post.
```
https://blog.replit.com/super-colliding-nix-stores

runeksvendsen commented:
Rendered draw.io diagram: 
![store-object-state-machine drawio](https://github.com/NixLayeredStore/rfcs/assets/1069247/f406471e-cdb9-4f85-aef0-547e50b1d873)

cbrewster commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,241 @@
+`LocalOverlayStore` is a subclass of `LocalStore` implementing the `local-overlay` store.
+It has additional configuration items for:
+
+ - The lower store, which must be a `LocalFSStore`
```
Does this preclude using special lower stores? (ie. the experimental FUSE-backed one)

benradf replied:
The FUSE-backed one should be okay, as long as it derives from `LocalFSStore`. This requirement just forces it to have a ["real store directory"](https://github.com/NixLayeredStore/nix/blob/ff12cf3b94e2303cb738f34712c0b8660f1261b6/src/libstore/local-fs-store.hh#L32-L41) that we can use as the `lowerdir` of the OverlayFS mount.

cbrewster commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,241 @@
+As discussed in the motivation, store objects are never knowingly duplicated.
+The `local-overlay` store while in operation ensures that store objects are stored exactly once:
+either in the lower store or the upper layer directory.
+No file system data should ever be duplicated by `local-overlay itself.
```
```suggestion
No file system data should ever be duplicated by `local-overlay` itself.
```

benradf commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,241 @@
+
+Add a new `local-overlay` store implementation to Nix.
+This will be a local store that is layered upon another local filesystem store (local store or daemon).
+This allows locally extending a shared store that is periodically extended with additional store objects.
```
Maybe:
> ... that is periodically _updated_ with additional store objects?

Since the two uses of 'extended' have overloaded the meaning of the word a little.

Ericson2314 replied:
```suggestion
This allows locally extending a shared store that is periodically updated with additional store objects.
```
Good idea!

benradf commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,241 @@
+
+Each has serious drawbacks:
+
+- Share nothing wastes tones of space as many duplicate store objects are stored separately.
```
_tons_ (or _tonnes_ for British spelling)

Might be worth mentioning it wastes time copying store objects around too.

benradf commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,241 @@
+  OverlayFS ensures that one copy in either store dir layer is enough, and we are careful never to wastefully include a store object in both layers.
+
+- No excess synchronization.
+  local changes are just that: local, not shared with any other consumer.
```
Nit: capitalization of _Local_

benradf commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,241 @@
+
+- No excess synchronization.
+  local changes are just that: local, not shared with any other consumer.
+  The lower store is never written to (no modifications or even filesystem locks) so any slow NFS write and sync paths should not be encountered.
```
> so any slow NFS write and sync paths should not be encountered.

I'm having a little trouble parsing this sentence properly. Maybe reorder it to be clearer; something like:

> so we should not encounter any slow NFS writes or path syncs.

benradf commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,241 @@
+Nix-unware usage can still work, but Nix-aware usage can do additional things.
+
+The `local-overlay` store can serve as a crucial tool to bridge these two modes of using Nix.
+The lower store can be however the artifacts were disseminated in the "hidden Nix" first phase of adoption, perhaps with a small tweak to expose the DB / daemon socket if it wasn't before.
```
Could the first part of the sentence be reworked a little? Initially I was really tripped up by trying to read 'however' as an isolated word, but I see now it's part of a bigger clause that follows:

> The lower store can be _however-the-artifacts-were-disseminated in the "hidden Nix" first phase of adoption_, ...

benradf commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,241 @@
+A slight drawback with the architecture is a lack of a normal form.
+A store object in the lower store may or may not have a DB entry in the `overlay-local` store.
+
+> In the store object state diagriam diagram, this is represented by the fact that there are two green nodes, instead of just one like the one blue node.
```
> In the store object state ~diagriam~ diagram

benradf commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,241 @@
+This is somewhat surprising, but reflects the fact that the lower store is logically immutable (even when it isn't a `local` store opened in read-only mode).
+By deleting the upper DB entry, we are not removing the object from the `local-overlay` store, but we are still resetting it to the initial state.
+
+# Alternatives
```
This seems like a good place to mention the `nix-store --load-db` alternative mentioned on the Super Colliding Nix Stores [discourse thread](https://discourse.nixos.org/t/super-colliding-nix-stores/28462/7). Downsides to that approach include:
- It takes 10 minutes to import the database for the 10TB store that Replit have.
- The database takes up 634MB (which is over half a user's quota).

Ericson2314 replied:
See what I wrote. I didn't see `nix-store --load-db` in the blog post unless I am missing something?

ryantm replied:
`nix-store --load-db` was suggested (on the NixOS Discourse linked above) as an alternative to our layered store approach here.

Ericson2314 replied:
Ah the *discourse* thread.

Ericson2314 replied:
Added something about that too.

benradf reviewed (commented):
Looks good! I added a couple of suggestions on content, but most of my comments are about minor grammatical things.

ryantm commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,263 @@
+
+- "Share everything" incurs major overhead from synchronization, even if consumers are making store objects they don't intend any other consumer to use.
+
+- Overlay everything cannot take advantage of new store objects added to the lower store, because its "fork" of the DB covers up the lower store's.
```
```suggestion
- Overlay everything cannot take advantage of new store objects added to the lower store, because its "fork" of the DB covers up the lower store's and could potentially lead to reading an incorrect SQLite Write-Ahead-Logging (WAL) file.
```

ryantm commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,263 @@
+The database is still a regular fresh empty one to start, and instead Nix explicitly knows about the lower store, so it can get any information for the DB it needs from it manually.
+This avoids all 3 downsides:
+
+- Store objects are never deduplicated.
```
```suggestion
- Store objects aren't duplicated.
```

ryantm commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,263 @@
+   This is specified with an escaped URL just like the `remote-store` setting of the two SSH stores types.
+
+ - The directory used as the upper layer of the OverlayFS
+
```
Might be nice to add an example here:
```suggestion
Here is an example of how the nix.conf configuration might look:
\```
store = local-overlay?lower-store=/mnt/lower/%3Fread-only%3Dtrue&upper-layer=/mnt/scratch/nix/upper/store
\```
```

ryantm commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,263 @@
+Any store object from the lower store that the `local-overlay` needs has that information copied into the `local-overlay` store's DB.
+This includes information for the closure of any such store object, because the normal closure property enforced by the DB's foreign key constraints is upheld.
+
+Store objects can still end up duplicated if the lower store gains a store object the upper store already had after.
```
```suggestion
Store objects can still end up duplicated if the lower store later gains a store object the upper store already had.
```

ryantm commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,263 @@
+
+> In the store object state diagram diagram, this is represented by the fact that there are two green nodes, instead of just one like the one blue node.
+
+This introduces some flexibility in the system: the seem "logical" layered store can be represented in multiple different "physical" configurations.
```
```suggestion
This introduces some flexibility in the system: the "logical" layered store can be represented in multiple different "physical" configurations.
```
was that supposed to be seemingly?

Ericson2314 replied:
Ah it was supposed to be "same"

Ericson2314 replied:
```suggestion
This introduces some flexibility in the system: the same "logical" layered store can be represented in multiple different "physical" configurations.
```

dfordivam commented on rfcs/0000-local-overlay-store.md:
````diff
@@ -0,0 +1,388 @@
+    D1("Lower layer filesystem data<br>Lower store metadata<br>Overlay DB metadata")
+```
+
+Whenever we the filesystem part of a store object reside in a layer, the metadata must also reside in that layer.
````
```suggestion
Whenever the filesystem part of a store object reside in a layer, the metadata must also reside in that layer.
```

dfordivam commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,388 @@
+(This is because the internal structure of the filesystem, unlike the internal structure of SQLite, is not visible to clients.)
+
+It is true that this is much slower used directly --- that is why Nix switched to using SQLite in the first place --- but in combination with a `local-overlay` store this doesn't matter.
+Since non-filesystem data is copied into the `local-overlay` store's DB, it will effectively act as a cache, speeding up future entires.
```
```suggestion
Since non-filesystem data is copied into the `local-overlay` store's DB, it will effectively act as a cache, speeding up future entries.
```

Ericson2314 replied:
Actually I think I meant this!
```suggestion
Since non-filesystem data is copied into the `local-overlay` store's DB, it will effectively act as a cache, speeding up future queries.
```

dfordivam commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,388 @@
+
+The downside of this is that Nix needs elevate permissions in order to create those bind mounts, and the impact of having arbitrarily many bind mounts is unknown.
+
+## Store implmenetations using FUSE
```
```suggestion
## Store implementations using FUSE
```

dfordivam commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,388 @@
+A slight drawback with the architecture is a lack of a normal form.
+A store object in the lower store may or may not have a DB entry in the `overlay-local` store.
+
+> In the store object state diagram diagram, this is represented by the fact that there are two green nodes, instead of just one like the one blue node.
```
```suggestion
> In the store object state diagram this is represented by the fact that there are two green nodes, instead of just one like the blue node.
```
Is this quoting someone?

Ericson2314 replied:
Oops that is about the previous version of the diagrams.

dfordivam commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,388 @@
+> In the store object state diagram diagram, this is represented by the fact that there are two green nodes, instead of just one like the one blue node.
+
+This introduces some flexibility in the system: the same "logical" layered store can be represented in multiple different "physical" configurations.
+This isn't a problem *per se*, but does mean there is a bit more complexity to consider during testing and system administration.
```
It would be good to have some elaborations on what the complexity increase will be.
Could a situation happen, where the user could end up "using" from the lower-layer, without having an entry in overlay db?

dfordivam replied:
If the system-admin does a mistake on remounting, and mounts a different lower layer, is there any way to detect it during init()? 
or things will blow up in runtime?
The answer to this may be obvious, but I dont know so just confirming

Ericson2314 replied:
> It would be good to have some elaborations on what the complexity increase will be.
Could a situation happen, where the user could end up "using" from the lower-layer, without having an entry in overlay db?

Part of my vagueness is that I suspect there is some ambiguity I haven't thought of.

> Could a situation happen, where the user could end up "using" from the lower-layer, without having an entry in overlay db?

That at least I don't think can happen.

Ericson2314 replied:
> If the system-admin does a mistake on remounting, and mounts a different lower layer, is there any way to detect it during init()?
> or things will blow up in runtime?

The "fsck" thing would catch it, but I fear it will be slow enough that it shouldn't be run unconditionally.

dfordivam replied:
> Could a situation happen, where the user could end up "using" from the lower-layer, without having an entry in overlay db?

Here I was trying to refer to a situation where a user does a wrong thing, like copy a dir from one system to another, which refer to some nix store paths.
Doing such a thing on other nix stores would cause it to fail, but it would likely "work" in this case. Or it might fail in unexpected manner, if these copied nix-store links are used to do some nix commands by the user.
This isn't a big problem, but it still is a case of non "normal" behaviour.

> The "fsck" thing would catch it, but I fear it will be slow enough that it shouldn't be run unconditionally.

This isn't an important issue at the moment, could be ignored..

dfordivam commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,388 @@
+
+We are considering a hybrid between the `local` store and `file://` store.
+This would *not* use NARs, but would use "NAR info" files instead of a SQLite database.
+This would side-step the currency issues of SQLite's read-only mode, and make "append only" usage of the store not require much synchronization.
```
```suggestion
This would side-step the concurrency issues of SQLite's read-only mode, and make "append only" usage of the store not require much synchronization.
```

dfordivam reviewed (commented):
Added some minor comments.
It may be good to capture any other scenarios to the lower layer, (other than the addition of objects, if they exist) and how those could be handled by the local-overlay.

And as for the future work, it may be necessary to add a functionality to move from a local-overlay to the regular local store. I believe the store "copy" operation would be relatively straightforward to implement.

benradf commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -385,3 +389,15 @@ This would side-step the concurrency issues of SQLite's read-only mode, and make
 Since non-filesystem data is copied into the `local-overlay` store's DB, it will effectively act as a cache, speeding up future queries.
 Each NAR info file only needs to be read once.
+
+## "Pivoting" betwen lower store
```
typo: _between_

Ericson2314 replied:
```suggestion
## "Pivoting" between lower store
```

benradf commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,403 @@
+### Both stores simplified view
+
+We can take the [Cartesian product](https://en.wikipedia.org/wiki/Cartesian_product_of_graphs) of these two graphs,
+and additionally tweak it to cover the exeption from above:
```
typo: _exception_

Ericson2314 replied:
```suggestion
and additionally tweak it to cover the exception from above:
```

cbrewster commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,403 @@
+
+- "Share nothing" wastes tons of space as many duplicate store objects are stored separately.
+
+- "Share everything" incurs major overhead from synchronization, even if consumers are making store objects they don't intend any other consumer to use.
```
I think there's also a security drawback here, consumers either need write permission in order to build things or they only have readonly permission and can't build new things. In our case we can't give users write access to shared resources.

cbrewster commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,403 @@
+    A -->|Create in overlay store| C
+
+    B -->|Create in upper store after created in lower store| D1
+    C -->|Create in lower store after created in lower store| D0
```
```suggestion
    C -->|Create in lower store after created in upper store| D0
```

cbrewster commented on rfcs/0000-local-overlay-store.md:
````diff
@@ -0,0 +1,403 @@
+```
+
+Whenever the filesystem part of a store object reside in a layer, the metadata must also reside in that layer.
+I.e. an lower layer of store dir store object must have lower store metadata (exact mechanism is abstract and unspecified, could be SQlite DB or daemon), and an upper layer of store object must have an overlay DB entry.
````
```suggestion
I.e. a lower layer of store dir store object must have lower store metadata (exact mechanism is abstract and unspecified, could be SQlite DB or daemon), and an upper layer of store object must have an overlay DB entry.
```

cbrewster commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,403 @@
+This gives a normal form in that objects are bind-mounted if and only if they have a DB entry in the bind-mounting store;
+There is no "have store object, don't yet have DB entry" middle state to worry about.
+
+The downside of this is that Nix needs elevate permissions in order to create those bind mounts, and the impact of having arbitrarily many bind mounts is unknown.
```
We also have to reconstruct repl container environments frequently, this would involve scanning the user's Nix SQLite DB and doing potentially hundreds or thousands of bind-mounts for each store entry at container start which is likely a non-starter

ryantm commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,403 @@
+
+2. The database can be very large.
+   For example, Replit's 16 TB store has a 634 MB database.
+   OverlayFS doesn't now how to duplicate only part of a SQLite database, so we have to duplicate the whole thing per user/consumer.
```
```suggestion
   OverlayFS doesn't know how to duplicate only part of a SQLite database, so we have to duplicate the whole thing per user/consumer.
```

ryantm commented on rfcs/0000-local-overlay-store.md:
```diff
@@ -0,0 +1,403 @@
+Running this periodically can allow the consumer to pick up any new paths added to the lower store since the last run.
+
+This does solve the first problem, but with poor performance penalty.
+We don't know what paths have change, so we have slurp up the entire DB.
```
```suggestion
We don't know what paths have changed, so we have slurp up the entire DB.
```

piegamesde committed:
Update the RFC process documentation (#150)

* Update the RFC process documentation

- Nudge authors towards using Semantic Line Breaks for writing the RFC text
- Nudge authors towards posting their text as pre-RFC in the forum first
- Tweak wording around who declares FCP slightly
- Make clear that the shepherds have to announce FCP, and when exactly the period officially starts
- Be more open about when to do the implementation work relative to the RFC, and reflect the current practice of starting implementation work early
- Make more clear that RFCs should not be amended, and that important information should live e.g. in documentation instead.

* RFC template: Use semantic line breaks

Given the short text this may not make a huge difference, but given
that every RFC starts with a copy of that template, the hope is that
it will nudge people to continue writing in that style.

* RFC template: Add "Prior art" section

This is mostly inspired by Rust's RFC template.
Also added a couple of words to the "Alternatives" section,
similarly inspired by Rust's template.

* fixup! Update the RFC process documentation

* fixup! Update the RFC process documentation

Co-authored-by: 7c6f434c <7c6f434c@mail.ru>

* fixup! Update the RFC process documentation

Co-authored-by: Valentin Gagarin <valentin.gagarin@tweag.io>

---------

Co-authored-by: 7c6f434c <7c6f434c@mail.ru>
Co-authored-by: Valentin Gagarin <valentin.gagarin@tweag.io>

Ericson2314 committed:
Mention that the read-only local store is already approved on an experimental basis

Ericson2314 committed:
Fill in some RFC metadata

Ericson2314 commented:
https://github.com/NixOS/rfcs/pull/152 This is now an official (proposed) RFC!
